### PR TITLE
Add CDS sources using the CLI

### PR DESCRIPTION
**Pull request by marikaner**
## Proposed Changes

Adds a command and init flag to enable the creation of a CAP compatible project.
Calling `sap-cloud-sdk init --addCds` in an empty directory will create a nest app with an additional module that exposes the CDS example catalogue service, as well as a `.cdsrc` and the`db` and `srv` directories. I also adds dependencies and scripts to the package.json and changes the `.gitignore`. The same command called within an exiting project will create the same resources except for the additional example nest module (as we don't know whether nest is used). `sap-cloud-sdk add-cds` again does the same but without the standard `init` parts.

Limitations: This does not consider deploying on CF. We have to clarify with the CAP colleagues what the best practice is here.

## Checklist

- [x] I have added or adjusted tests that prove my fix is effective or that my feature works
- [x] I have added or adjusted documentation

## Further Comments

If the changes are complex, please discuss what alternatives you have evaluated and why you picked this particular solution. Please make sure to highlight any trade-offs that we should know about.

### CODE REVIEW

File: src/utils/package-json.ts [hunk 1]
```diff
@@ -101,9 +115,29 @@ function mergePackageJson(originalPackageJson: any, changes: any) {
...
+
+export async function modifyPackageJson({
+  projectDir,
+  isScaffold = false,
+  frontendScripts = false,
+  force = false,
+  addCds = false
+}: {
+  projectDir: string;
+  isScaffold?: boolean;
```
Comment (florian-richter) [inline src/utils/package-json.ts]: This interface is almost `ModifyPackageJsonParams` why not use that?

File: src/utils/scaffold.ts [hunk 2]
```diff
@@ -47,14 +47,36 @@ export async function buildScaffold(projectDir: string, verbose: boolean) {
...
+    addCatalogueModule(path.resolve(projectDir, 'src', 'app.module.ts'));
+  }
+  cli.action.stop();
+}
+
+function modifyMainTs(pathToMainTs: string) {
   const mainTs = fs.readFileSync(pathToMainTs, { encoding: 'utf8' });
-  const modifiedMainTs = mainTs.replace('.listen(3000)', '.listen(process.env.PORT || 3000)');
+  const modifiedListen = '.listen(process.env.PORT || 3000)';
+  const modifiedMainTs = mainTs.replace('.listen(3000)', modifiedListen);
```
Comment (florian-richter) [inline src/utils/scaffold.ts]: This seems unnecessary

File: test/add-approuter.spec.ts [hunk 3]
```diff
@@ -36,9 +37,7 @@ describe('Add Approuter', () => {
 
   it('should add preconfigured files', async () => {
     const projectDir = path.resolve(pathPrefix, 'add-approuter');
-    if (fs.existsSync(projectDir)) {
-      fs.removeSync(projectDir);
-    }
+    removeDir(projectDir);
```
Comment (florian-richter) [inline test/add-approuter.spec.ts]: This could use the same function used in other test files: `const projectDir = getCleanProjectDir(pathPrefix, 'add-approuter');`

File: test/add-cx-server.spec.ts [hunk 4]
```diff
@@ -18,11 +18,13 @@ jest.mock('cli-ux', () => {
...
 import * as path from 'path';
 import AddCxServer from '../src/commands/add-cx-server';
+import { getPathPrefix, removeDir } from './test-utils';
 
 describe('Add CX Server', () => {
-  const pathPrefix = path.resolve(__dirname, __filename.replace(/\./g, '-')).replace('-ts', '');
+  const pathPrefix = getPathPrefix(__dirname, __filename);
 
   beforeAll(() => {
+    // TODO: remove
```
Comment (florian-richter) [inline test/add-cx-server.spec.ts]: TODO

File: package.json [hunk 5]
```diff
@@ -36,6 +39,7 @@
     "fs-extra": "^8.1.0",
     "globby": "^10.0.1",
     "jest": "^24.9.0",
+    "jest-extended": "^0.11.2",
```
Comment (florian-richter) [inline package.json]: What is this used for?

Comment (marikaner) [reply package.json]: Simplified testing. e. g. `expect(...).toIncludeAllMembers`, `expect(...). toStartWith`, `expect(...). toEndWith`.

File: src/commands/add-approuter.ts [hunk 6]
```diff
@@ -17,24 +17,32 @@ export default class AddApprouter extends Command {
...
     }),
     force: flags.boolean({
       description: 'Do not fail if a file already exist and overwrite it.'
     }),
     help: flags.help({ char: 'h' })
   };
 
+  static args = [
+    {
+      name: 'projectDir',
```
[the remark this answers is not in the document]
Comment (marikaner) [reply src/commands/add-approuter.ts]: But is it an issue? I would argue either we have it or we don't. So either remove the flag or keep both for consistency (and make it non-hidden, I guess). What do you think?

File: src/commands/add-cds.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+Generated service needs to be exposed.
+For express apps you can do this by adding the following snippet to your code:
+  cds
+    .connect()
+    .serve('CatalogService')
+    .in(<your-express-app>)
+For other frameworks please refer to the documentation.
+`
+    );
```
[the remark this answers is not in the document]
Comment (marikaner) [reply src/commands/add-cds.ts]: Hmm, do you have a different proposal?

File: src/commands/init.ts [hunk 8]
```diff
@@ -109,7 +106,14 @@ export default class Init extends Command {
         {
           title: 'Creating files',
           task: () => {
-            const copyDescriptors = getCopyDescriptors(projectDir, getTemplatePaths(['init']));
+            const templates = ['init'];
+            if (flags.addCds) {
+              templates.push('add-cds');
+              if (isScaffold) {
+                templates.push('add-cds-scaffold');
```
[the remark this answers is not in the document]
Comment (marikaner) [reply src/commands/init.ts]: yeah, it's ugly. 😊

File: src/commands/add-cds.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (florian-richter) [reply src/commands/add-cds.ts]: I prefer
```ts
this.log('line1');
this.log('line2');
this.log('line3');
```

It's not the prettiest solution but I find it easier to not mess up this way.

File: src/commands/add-approuter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (florian-richter) [reply src/commands/add-approuter.ts]: My assumption was that a simpler (public) API makes it easier to understand the docs. So omitting a flag/arg that is not useful for 99% of users (and can be simply worked around by `cd projectDir`) would be a positive change.

But I have not validated that assumption, so feel free to change my mind on this.

File: src/templates/add-cds-scaffold/src/catalogue/catalogue-service-handler.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Implementation for CatalogService defined in ./cat-service.cds
+ */
+import cds = require('@sap/cds');
+
+export const serviceHandler = srv => {
+  // Use reflection to get the csn definition of Books
```
[the remark this answers is not in the document]
Comment (marikaner) [reply src/templates/add-cds-scaffold/src/catalogue/catalogue-service-handler.ts]: **C**ore **S**ervice **N**otation, but I will change this file and add a comment where to use the sdk.

File: src/templates/add-cds/db/data/my.bookshop-Books.csv [hunk 10]
```diff
@@ -0,0 +1,6 @@
+ID;title;author_ID;stock
+201;Wuthering Heights;101;12
+207;Jane Eyre;107;11
+251;The Raven;150;333
+252;Eleonora;150;555
```
[the remark this answers is not in the document]
Comment (marikaner) [reply src/templates/add-cds/db/data/my.bookshop-Books.csv]: 😂

File: src/utils/manifest-yaml.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+
+export function getProjectNameFromManifest(command: Command): string | undefined {
+  try {
+    const manifestStr = fs.readFileSync('manifest.yml', { encoding: 'utf8' });
+    const manifest = yaml.safeLoad(manifestStr, {
+      filename: 'manifest.yml',
+      onWarning: () =>
+        command.warn('Could not read name from `manifest.yml`. Please ensure you ran `sap-cloud-sdk init` before adding the approuter.') // TODO:
```
[the remark this answers is not in the document]
Comment (marikaner) [reply src/utils/manifest-yaml.ts]: ✅

File: src/utils/package-json.ts (hunk 1 above)
Comment (marikaner) [reply src/utils/package-json.ts]: I wanted to delete it, just forgot to do it. ;)

File: src/commands/add-approuter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (marikaner) [reply src/commands/add-approuter.ts]: I have removed all the `--projectDir` flags as discussed, but I decided to leave the argument, because I think it makes our commands consistent. I don't think `projectDir` makes the API hard to understand, it is still very simple, while the change is tiny.

File: src/utils/scaffold.ts (hunk 2 above)
Comment (marikaner) [reply src/utils/scaffold.ts]: I allows you to not log a warning if you already had this in your code (which might happen if this gets changed in nest)

File: src/commands/add-approuter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mr-flannery) [reply src/commands/add-approuter.ts]: To throw my 2cts in, I also have no idea whether we need it or not. So the question is how do find that out in the future? Otherwise we'll have the same discussion on the next PR.

File: src/utils/package-json.ts [hunk 12]
```diff
@@ -64,12 +74,16 @@ function findScriptConflicts(originalScripts: any, scriptsToBeAdded: any) {
...
 }
 
-async function getPackageJsonChanges(isScaffold: boolean, frontendScripts: boolean) {
+async function getPackageJsonChanges(isScaffold: boolean, frontendScripts: boolean, addCds: boolean) {
   const changes: PackageJsonChange[] = [isScaffold ? scaffoldProjectPackageJson : existingProjectPackageJson];
   if (frontendScripts) {
     changes.push(frontendTestScripts);
   }
 
+  if (addCds) {
```
Comment (mr-flannery) [inline src/utils/package-json.ts]: As for printing the message, something declarative might be a good idea here. It's probably too early to generalize, but given that there are things to add to the package.json for each thing that is added to the project, it might be a good idea to hold a mapping of `thing` to `packageJsonAdditions` somewhere and then just collect, merge and write it.
But we should probably wait a little longer here before generalizing.

File: src/utils/scaffold.ts [hunk 13]
```diff
@@ -47,14 +47,36 @@ export async function buildScaffold(projectDir: string, verbose: boolean) {
...
+  if (addCds) {
+    addCatalogueModule(path.resolve(projectDir, 'src', 'app.module.ts'));
+  }
+  cli.action.stop();
+}
+
+function modifyMainTs(pathToMainTs: string) {
   const mainTs = fs.readFileSync(pathToMainTs, { encoding: 'utf8' });
-  const modifiedMainTs = mainTs.replace('.listen(3000)', '.listen(process.env.PORT || 3000)');
+  const modifiedListen = '.listen(process.env.PORT || 3000)';
```
Comment (mr-flannery) [inline src/utils/scaffold.ts]: Since we're replacing code here anyway, what do you think of changing the default port to 8080? You need to change it anyway if you want to deploy on CF. (I'm aware that this is out of scope for this PR)

File: test/test-utils.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+ */
+import * as fs from 'fs';
+import * as path from 'path';
+import * as rm from 'rimraf';
+
+export function getPathPrefix(dir: string, file: string): string {
+  return path.resolve(dir, file.replace(/\./g, '-')).replace('-ts', '');
+}
+
+export function getCleanProjectDir(pathPrefix: string, name: string): string {
```
Comment (mr-flannery) [inline test/test-utils.ts]: 👍

File: test/test-utils.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
+/*!
+ * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
+ */
+import * as fs from 'fs';
+import * as path from 'path';
+import * as rm from 'rimraf';
+
+export function getPathPrefix(dir: string, file: string): string {
```
Comment (mr-flannery) [inline test/test-utils.ts]: I think the name of the function could be improved. If I understood it correctly, this function gives you a path under which a folder is created in which a test can write stuff isolated from different tests.
Assuming my understanding is correct, maybe a more specific name like `testOutputPath` or `testOutputPathFolder` can help to convey the use of this function.

File: test/utils/scaffold.spec.ts [hunk 16]
```diff
@@ -47,10 +39,10 @@ describe('Scaffold Utils', () => {
...
   it('should build the scaffold', async () => {
-    const projectDir = getCleanProjectDir('build-scaffold');
+    const projectDir = getCleanProjectDir(pathPrefix, 'build-scaffold');
 
-    await buildScaffold(projectDir, false);
+    await buildScaffold(projectDir, false, false);
 
     expect(fs.readdirSync(projectDir).sort()).toMatchSnapshot();
-  }, 60000);
+  }, 80000);
```
Comment (mr-flannery) [inline test/utils/scaffold.spec.ts]: Meh. We should probably look into a way soon to not have to increase test timouts regularly...

File: test/utils/templates.spec.ts [hunk 17]
```diff
@@ -31,19 +23,19 @@ describe('Templates Utils', () => {
...
   });
 
   it('should copy files locally', async () => {
-    const projectDir = getCleanProjectDir('copy-files-locally');
+    const projectDir = getCleanProjectDir(pathPrefix, 'copy-files-locally');
     copyFiles(getCopyDescriptors(projectDir, getTemplatePaths(['init'])), {});
     expect(fs.readdirSync(projectDir).sort()).toMatchSnapshot();
   });
 
   // it('should copy files remotely', async () => {
```
Comment (mr-flannery) [inline test/utils/templates.spec.ts]: commented code

File: src/utils/git-ignore.ts [hunk 18]
```diff
@@ -27,3 +30,7 @@ export function modifyGitIgnore(projectDir: string) {
     pathsToIgnore.forEach(path => cli.log('  ' + path));
   }
 }
+
+function cdsPathsToIgnore() {
```
Comment (mr-flannery) [inline src/utils/git-ignore.ts]: This could be a `const` instead of a `function` (unless you're concerned that someone ninja pushes values into it).

File: test/utils/scaffold.spec.ts (hunk 16 above)
Comment (marikaner) [reply test/utils/scaffold.spec.ts]: +1
